test/extensions/filters/http/dynamic_forward_proxy/proxy_filter_test.cc

```diff
@@ -475,6 +475,50 @@ TEST_F(UpstreamResolvedHostFilterStateHelper, UpdateResolvedHostFilterStateMetad
   filter_->onDestroy();
 }
 
+// Tests if address set is populated in the filter state when an upstream host is resolved
+// successfully but is null.
```

Comment:
Sorry I don't understand this case. How can there be an address that is null? I can see there being no address but I don't think we should ever see null? What's the case in which this fails?

/wait-any

Reply:
> Sorry I don't understand this case. How can there be an address that is null? I can see there being no address but I don't think we should ever see null? What's the case in which this fails?
> 
> /wait-any

This was to add coverage for https://github.com/envoyproxy/envoy/blob/3f3577ddc6719109d3c33ca22c712348f898b020/source/extensions/filters/http/dynamic_forward_proxy/proxy_filter.cc#L159.

Reply:
But why do you have that case? I don't understand how it happens.

/wait

Reply:
I could be getting this completely wrong but one way i can make it happen is when I run v6 tests without a v6 interface. In this case, the DNS resolver notifies the threads (dns_cache_impl.cc:348) even when the resolved address is still null.

Reply:
When a resolution comes back with no address, on the first resolve, it will unblock all the threads. I would need to look at the code in more detail in terms of how this is exposed, but that I believe is what is happening. There should already be test cases for this so you shouldn't need a new test case? Can you make sure you look at the existing failed / no host resolved test cases and see what is going on there?

Reply:
To clarify - when you mean `no address`, it is a null address? If so, when the threads are unblocked, `ProxyFilter::onLoadDnsCacheComplete` is called with a null resolved address.

Reply:
I would need to go back to the code to see what it does but basically this is the flow:
1) Start resolve.
2) If first resolve fails or returns no addresses, unblock threads.
3) Threads I think will find no healthy host to route to, but I don't think they should find no address.

Please look at the code and see how this is happening and there should be a test that already covers this case. I don't think you need a new test case. If you do I don't understand what case you are hitting and you need to debug it.

Reply:
This is my understanding:

Step #3  for DFP split into :
3.1. Add the resolved address to FilterState (this PR)
3.2 Continue with normal routing (no healthy host)

Since this PR introduces 3.1 and we dont want to add a null address in the filter state, i had to add a check.

Reply:
Yes but why do you need a new test? This case should already be covered.

Reply:
The coverage tests fails since we don't have tests that cover the branch of code https://github.com/envoyproxy/envoy/blob/3f3577ddc6719109d3c33ca22c712348f898b020/source/extensions/filters/http/dynamic_forward_proxy/proxy_filter.cc#L159.

Reply:
OK sorry I don't have any further time to help you right now, but I would like to understand this more. Something doesn't make sense here. The case of no addresses should already be covered. Please look into it further.

Reply:
Can i just remove the check for null address?  This will expose the FilterState to have null address.  If I leave the check as it is, the overall system would work ok due to other checks in place but coverage tests will fail. The test was added just to get CI passing.

Reply:
I don't think it makes sense to add a null address to the set. Please debug it and understand what is going on. Thanks!

Reply:
In `ProxyFilter::onLoadDnsCacheComplete` (the function that is called to notify on first resolution complete) before this PR, the `DnsHostInfoSharedPtr` parameter isn't used/examined in any way. With this PR, it is now used, to get the initial address. That's why this wasn't covered previously. But it may make sense to move the nullptr check into `onLoadDnsCacheComplete()` instead of where it is now; I don't think there should be a null address from the other path.

Reply:
Yup that makes sense, thanks.

Reply:
yeah we should probably fix updates in this case.  I was a bit worried it'd cause problems later in the chain but https://github.com/envoyproxy/envoy/pull/18150 definitely has a null address (I added then removed assert checks) and just thankfully fails to connect rather than trying to null ptr reference.

Reply:
Wanted to clarify - i will leave the check for `address` as it is since we need the check only when adding the address to the FilterState.